src/transformers/testing_utils.py

```diff
@@ -1615,3 +1617,32 @@ def new_request(self, method, **kwargs):
             self.other_request_count += 1
 
         return self.old_request(method=method, **kwargs)
+
+
+# adapted from https://stackoverflow.com/questions/32163436/python-decorator-for-printing-every-line-executed-by-a-function
+class set_reproducible(ContextDecorator):
+    r"""
+    A context decorator to set a manual seed before each PyTorch's forward pass. This is useful for models that
+    involves stochasticity during their forward pass to ensure reproducibility and the capability to run som tests.
+    """
+
+    def __init__(self, seed=42):
+        super().__init__()
+        self.seed = seed
+
+    def __enter__(self):
+        sys.settrace(self.trace_calls)
+
+    def __exit__(self, *exc):
+        # stop tracing functions after quitting
+        # the context manager
+        sys.settrace(None)
+
+    def trace_calls(self, frame, event, arg):
+        # Set the seed when it is a call to a forward function from the model
+        if "/modeling_" in frame.f_code.co_filename:
+            return self.set_seed
```

Comment:
So if I understand it correctly, you're using this hack to force the same seed before every python call, correct?

What is it that you're testing then, since this is not the normal behavior, i.e. normal code base sets the seed once to be reproducible.

In other words why setting the seed at the beginning of the program doesn't lead to a reproducible outcome? Isn't that an indicator of a bug in the proposed software and this hack tries to hide it?

and if you're really going to keep this can we call it something with a better mnemonic like `with reset_seed_on_every_frame`? Since `set_reproducible` to me means setting a seed once at the enter and leaving it alone.

Reply:
Thanks @stas00 for your reply! 

Regarding your first point yes but, we set the seed before each `forward` call from a Pytorch module instead of any python call (from @ArthurZucker 's modification).

Regarding your second point, I was also very confused at the beginning but setting the pytorch seed before each `forward` seems to be needed to ensure reproducibility for stochastic operations. E.g. this code snippet shows that:
```
>>> import torch
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x10576b6f0>
>>> torch.randn(1, 3)
tensor([[ 1.5410, -0.2934, -2.1788]])
>>> torch.randn(1, 3)
tensor([[ 0.5684, -1.0845, -1.3986]])
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x10576b6f0>
>>> torch.randn(1, 3)
tensor([[ 1.5410, -0.2934, -2.1788]])
```
You need to set the seed before each stochastic call to ensure reproducibility. From my understanding that is why we need to call the seed setting on stochastic tests (e.g. [here](https://github.com/huggingface/transformers/blob/0f257a87749e0a72bda260c6f319a45dae1e7c4d/tests/models/vit_mae/test_modeling_vit_mae.py#L235) where the seed is called before each forward pass. I am still not sure why calling the seed once is not sufficient but I think that it might be because the seed has been 'consumed' when a stochastic operation is performed (This has been also observed on Jukebox integration I think).

For the last point yes! I think that we should change the decorator's name to make it more understandable

Reply:
Ah, I missed that nuance. Then `reset_seed_before_every_forward` would be more appropriate.

But there is already a mechanism for doing this - this is called `register_forward_hook`
https://pytorch.org/docs/stable/generated/torch.jit.ScriptModule.html?highlight=hook#torch.jit.ScriptModule.register_forward_hook

so you don't need any cool hacks to accomplish that.

e.g. see:

https://github.com/huggingface/transformers/blob/4c2e983f44ce4d3b9c8502d42cc568e45897bd15/src/transformers/debug_utils.py#L160

Reply:
Very interesting, both option work, and maybe `register_forward_pre_hook(hook)` is more appropriate. 
IMO having a decorator will allow anyone to use it without modifying their test function, but we might have a lot of decorators. 

We could define the following
```python
# testing_utils.py
def pre_hook_set_seed(seed):
    def pre_hook(module, input):
        set_seed(seed)
    return pre_hook
```
And call 
```python 
# test_modeling_vilt.py
handle = self.register_forward_pre_hook(pre_hook_set_seed)
...
handle.remove()
```
I guess its pretty similar but requires a bit more code. 😄 
@stas00 What do you think would work best long term?

Reply:
the pytorch call is designated for this specific functionality so it's a much cleaner solution, IMHO.

You can just as well implement it as a container so enter/exit for the above sample.

Also I was thinking that if ViLT needs this functionality often, perhaps the hook should be added directly into its modeling file?

But testing_utils is perfect too if it's generic and other models might need it too.

Reply:
Agreed! I will try to make a decorator using the `ur f`. 
Thanks for your feedbacks

Reply:
Hey! So, the problem with the pytorch hook is that you need to grab the module instance in order to link the hook. Thus you can't really just have a decorator for the test, you need to re-write the test (as the models are accessed by their names), or maybe I was just unable to do so 😅 . 

Here is the final solution : 

```python 
    def trace_calls(self, frame, event, arg):
        # frame.f_locals provides the arguments of the function that was called 
        # we just check that it is a ...Model class 
        with suppress(KeyError):
            if "Model" in str(frame.f_locals["self"].__class__):
                return self.set_seed
``` 
@sgugger  What do you think?

Reply:
This seems a bit overengineered for three tests where we could just insert two `set_seed` calls to ensure the same results (even registering a pre-foward hook seems over-engineered TBH). Or am I missing something?

Reply:
You are right 😅 But since `VitMae` also requires that hack, we though that long term, made more sense to have an easy to use decorator! You have to insert the `set_seed` before every `forward` (so in the for loop that is testing) for each of the test so the original idea to have a decorator followed Yih-dar's comment. 
Anyway will do as you think works better!

Reply:
While I was excited by the idea at first, reading Stas' comments and thinking more about it, I think we should stick to something simple :-)

Reply:
Why does it have to be a test decorator? what would be wrong with calling it from the test?

Also I was thinking that if this might be used by users - e.g. to debug their code, won't it be a good idea to make this functionality as part of the model itself?

So basically one would call:

```
model = from_pretrained(...)
model.reset_seed_before_every_forward(True)
``` 

and so now it's accessible to tests and users too.

If multiple models require that it could be a mixin class and they can inherit from it, so it won't make any noise in the modeling file.

Reply:
Also fine with this.